reolink_aio/typings.py

```diff
@@ -27,6 +30,177 @@
+    
+    @property
+    def hour(self)->int:
+        """rule hour"""
+        return self.data[f"{self.prefix}Hour"]
+
+    @property
+    def minute(self)->int:
+        """rule minute"""
+        return self.data[f"{self.prefix}Min"]
+
+    @property
+    def second(self)->int:
+        """rule second"""
+        return self.data[f"{self.prefix}Sec"]
+
+    def time(self)->time:
+        """rule as time"""
+        return time(self.hour, self.minute, self.second)
+    
+    def date(self, year:int)->date:
+        """rule as date with year"""
+        weekday =  self.weekday
+        # 5 is the last week of the month, not necessarily the 5th week
+        if self.week == 5:
+            if self.month < 12:
+                month = self.month + 1
+            else:
+                month = 1
+                year += 1
+            __date = date(year, month, 1)
+            if __date.weekday() < weekday:
+                __date -= timedelta(weeks=1)
+            __date -= timedelta(days=__date.weekday() - weekday)
+        else:
+            __date = date(year, self.month, 1) + timedelta(weeks=self.week)
+            __date += timedelta(days=__date.weekday() - weekday)
+        return __date
+
+    def datetime(self, year:int)->datetime:
+        """rule as datetime with year"""
+        return datetime.combine(self.date(year), self.time())
+
+class _YearStartStopCache(dict[str,tuple[datetime,datetime]]):
+    def __init__(self, factory:Callable[[int], tuple[datetime,datetime]]):
+        super().__init__()
+        self.factory = factory
+
+    def __missing__(self, key:int)->tuple[datetime,datetime]:
+        self[key] = self.factory(key)
+        return self[key]
+
+
+class Reolink_timezone(tzinfo):
+    """Reolink DST timezone implementation"""
+
+    _cache:ClassVar[dict[tuple,"Reolink_timezone"]] = {}
+
+    @classmethod
+    def get(cls, data:dict[str,Any])->"Reolink_timezone":
+        """get or create timzeone based on data"""
+        if TYPE_CHECKING:
+            __data:GetTimeResponse = data
+        else:
+            __data = data
+
+        key = tuple([__data["Time"]["timeZone"]]) + tuple(__data["Dst"].values())
+        return cls._cache.setdefault(key, cls(data))
+
+    def __init__(self, data:dict[str,Any]) -> None:
+        super().__init__()
+        if TYPE_CHECKING:
+            __data = cast(GetTimeResponse, data)
+        else:
+            __data = data
+        self._dst = timedelta(hours=__data["Dst"]["offset"]) if bool(__data["Dst"]["enable"]) else timedelta(hours=0)
+        # Reolink does a positive UTC offset but python expects a negative one
+        self._offset = timedelta(seconds=-__data["Time"]["timeZone"])
+        start_rule = _DST_Rule(__data["Dst"], "start")
+        end_rule = _DST_Rule(__data["Dst"], "end")
+        self._year_cache = _YearStartStopCache(lambda year: (start_rule.datetime(year), end_rule.datetime(year)))
+
+    def tzname(self, __dt: datetime | None) -> str | None:
```

Comment:
Why add it if we do not have it?

Reply:
I am assuming you mean why the year cache....
The rules are calculations, find the start/end dates of DST for the given year. we cache it so we only do the calculation once, since it will never change for that ruleset.

Reply:
tzname, utcoffset, and dst are required to have a definition when implemention the tzinfo class. None is an allowable return from tzname.

(https://docs.python.org/3/library/datetime.html#tzinfo-objects)